File: companion/ios/CanRosettaCompanion/IO/JSONLWriter.swift
import Foundation
import os

/// A buffered, thread-safe writer that serialises `Encodable` records one per
/// line (newline-delimited JSON, `.jsonl`) to a file.
///
/// Used for `phone/motion.jsonl`, `phone/location.jsonl` and
/// `phone/video_index.jsonl`. All appends are funnelled through a private
/// serial queue, so it is safe to call `append` from the CoreMotion operation
/// queue, the CoreLocation delegate queue and the capture queue concurrently.
///
/// Records are encoded with `.convertToSnakeCase`, so idiomatic camelCase Swift
/// property names (`tUtc`, `hAcc`) map to the wire names the schemas require
/// (`t_utc`, `h_acc`).
final class JSONLWriter {

    private let handle: FileHandle
    private let queue: DispatchQueue
    private let encoder: JSONEncoder
    private let logger = Logger(subsystem: AppInfo.subsystem, category: "jsonl")

    /// Bytes buffered before a flush to disk. IMU at 100 Hz produces small
    /// lines; flushing in ~64 KiB chunks keeps syscall overhead low without
    /// risking much data on a crash.
    private let flushThreshold = 64 * 1024
    private var buffer = Data()

    /// Number of records written so far. Read via `rowCount` (thread-safe).
    private var count = 0

    let url: URL

    init(url: URL) throws {
        self.url = url
        FileManager.default.createFile(atPath: url.path, contents: nil)
        self.handle = try FileHandle(forWritingTo: url)
        self.queue = DispatchQueue(label: "\(AppInfo.subsystem).jsonl.\(url.lastPathComponent)")

        let enc = JSONEncoder()
        enc.keyEncodingStrategy = .convertToSnakeCase
        enc.outputFormatting = [.withoutEscapingSlashes] // compact, one line
        self.encoder = enc
    }

    /// Append one record. Non-blocking: encoding and buffering happen on the
    /// serial queue. Encoding failures are logged and the record is dropped
    /// (we never want a bad sample to abort a drive-long recording).
    func append<T: Encodable>(_ record: T) {
        queue.async { [weak self] in
            guard let self else { return }
            do {
                var line = try self.encoder.encode(record)
                line.append(0x0A) // '\n'
                self.buffer.append(line)
                self.count += 1
                if self.buffer.count >= self.flushThreshold {
                    self.flushLocked()
                }
            } catch {
                self.logger.error("Failed to encode record: \(error.localizedDescription)")
            }
        }
    }

    /// Thread-safe row count for the manifest / live UI.
    var rowCount: Int {
        queue.sync { count }
    }

    /// Flush remaining buffer and close the file. Blocks until done.
    func close() {
        queue.sync {
            flushLocked()
            do {
                try handle.close()
            } catch {
                logger.error("Failed to close \(self.url.lastPathComponent): \(error.localizedDescription)")
            }
        }
    }

    /// Must be called on `queue`.
    private func flushLocked() {
        guard !buffer.isEmpty else { return }
        do {
            try handle.write(contentsOf: buffer)
            buffer.removeAll(keepingCapacity: true)
        } catch {
            logger.error("Failed to write to \(self.url.lastPathComponent): \(error.localizedDescription)")
        }
    }
}

File: companion/android/app/src/main/java/com/inomotech/canrosetta/companion/io/JsonlWriter.kt
package com.inomotech.canrosetta.companion.io

import android.util.Log
import com.inomotech.canrosetta.companion.AppInfo
import org.json.JSONObject
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicLong

/**
 * A buffered, thread-safe writer that serialises records one per line
 * (newline-delimited JSON) to a file. All disk writes are funnelled through a
 * private single-thread executor, so it is safe to call [append] concurrently
 * from the sensor handler thread, the location callback and the camera analyzer.
 *
 * Used for `phone/motion.jsonl`, `phone/location.jsonl`, `phone/video_index.jsonl`
 * and `phone/photos_index.jsonl`.
 */
class JsonlWriter(file: File) {

    private val out = BufferedOutputStream(FileOutputStream(file))
    private val executor: ExecutorService = Executors.newSingleThreadExecutor()
    private val counter = AtomicLong(0)

    @Volatile
    private var closed = false

    val fileName: String = file.name

    /**
     * Append one record. The count is bumped synchronously (so [rowCount] is
     * immediately accurate for the manifest / UI); encoding and the disk write
     * happen on the writer thread. Write failures are logged and the record is
     * dropped — a bad sample must never abort a drive-long recording.
     */
    fun append(obj: JSONObject) {
        if (closed) return
        counter.incrementAndGet()
        val bytes = (obj.toString() + "\n").toByteArray(Charsets.UTF_8)
        executor.execute {
            try {
                out.write(bytes)
            } catch (e: IOException) {
                Log.e(AppInfo.TAG, "jsonl write failed ($fileName): ${e.message}")
            }
        }
    }

    /** Thread-safe row count. */
    val rowCount: Long get() = counter.get()

    /** Flush the remaining buffer and close the file. Blocks until done. */
    fun close() {
        if (closed) return
        closed = true
        executor.execute {
            try {
                out.flush()
                out.close()
            } catch (e: IOException) {
                Log.e(AppInfo.TAG, "jsonl close failed ($fileName): ${e.message}")
            }
        }
        executor.shutdown()
        try {
            executor.awaitTermination(10, TimeUnit.SECONDS)
        } catch (e: InterruptedException) {
            Thread.currentThread().interrupt()
        }
    }
}
